File: js/render/viewmodel.js
// The first-person hand: how the held item is posed, framed and animated.
//
// Everything about the viewmodel lives here so that adding an item never means
// touching the renderer:
//
//  • STYLES   — a table of hold styles, pure data. Where the grip sits on the
//               screen, how the item is turned in the hand, and how big it
//               reads. Add a style, or tweak one, without reading a line of GL.
//  • styleFor — resolves an item to a style. An item can name one explicitly
//               with `hold: "..."` in the registry; otherwise it falls out of
//               its type (tool / sword / food / block) so new items inherit a
//               sensible pose for free.
//  • Viewmodel — the animation clocks (swing, equip, walk sway, idle drift).
//               They only ever nudge the same numbers the style defines, so an
//               animation can never disagree with a pose.
//
// The framing is deliberately resolution-independent: `anchor` is a fraction of
// the half-frame (±1 = screen edge) and `size` is a fraction of the frame
// HEIGHT, both measured at the hold distance. So the item sits in the same spot
// of the picture on an ultrawide monitor, a phone, or a 4:3 window — the old
// fixed view-space offsets slid off the edge as the aspect ratio narrowed.

// Every style is complete (no inheritance to trace):
//   rot    [x, y, z]  radians, applied roll(z) -> pitch(x) -> yaw(y) about the grip.
//                     +z rolls the item's top to the LEFT, -x tips its top AWAY
//                     into the scene, -y turns its face back toward the camera.
//   anchor [x, y]     where the grip lands, as a fraction of the half-frame.
//                     [0,0] is the crosshair, [1,-1] the bottom-right corner.
//   size              model height as a fraction of the frame height. The mesh
//                     is unit-scaled (1 = one block, or one full 16px sprite),
//                     so a slab really does read as half a block.
//   dist              how far in front of the camera the item hangs.
//   grip   [x, y, z]  the model-space point that lands on the anchor. Item
//                     meshes are built bottom-centred, so [0,0,0] is the butt of
//                     a tool's handle / the bottom of a block.
//   flip              mirror the model's x. Sword art is drawn hilt-low-left,
//                     blade-high-right (the icon convention); mirrored, the hilt
//                     falls into the hand and the blade reaches for the crosshair.
export const STYLES = {
  // A block, held out at the classic three-quarter angle: you can see its top
  // and two sides at once, so stairs read as stairs and slabs as slabs.
  block: {
    rot: [0.30, -0.66, 0], anchor: [0.66, -0.95], size: 0.34, dist: 0.62,
    grip: [0, 0, 0], flip: false,
  },
  // Pick / axe / shovel: handle butt in the hand at the bottom right, rolled so
  // the head leans up toward the crosshair, tipped away so it reaches into the
  // scene, and turned a few degrees off face-on so the extruded edge shows and
  // the sprite reads as an object rather than a sticker.
  tool: {
    rot: [-0.28, -0.46, 0.52], anchor: [0.62, -1.02], size: 0.56, dist: 0.62,
    grip: [0, 0, 0], flip: false,
  },
  // A sword stands closer to upright than a tool. Its art is drawn on the
  // diagonal, so the roll here is NEGATIVE — it cancels most of the 45° the
  // sprite already has and lifts the blade up the screen.
  sword: {
    rot: [-0.22, -0.46, -0.48], anchor: [0.62, -1.06], size: 0.58, dist: 0.62,
    grip: [0.30, 0.06, 0], flip: true,
  },
  // Food is brought up nearer the middle, turned to face you — you are about to
  // eat it, so it should be readable.
  food: {
    rot: [-0.10, -0.34, 0.22], anchor: [0.58, -0.92], size: 0.40, dist: 0.62,
    grip: [0, 0, 0], flip: false,
  },
  // Doors, trapdoors and ladders are flat panels: held at a cube's angle they
  // are a meaningless sliver, so they get turned to show their face.
  panel: {
    rot: [0.10, -1.15, 0.18], anchor: [0.66, -0.92], size: 0.44, dist: 0.62,
    grip: [0, 0, 0], flip: false,
  },
  // Everything else: ingots, gems, buckets, torches, the atlas. Mostly face-on
  // with a hint of turn so the extrusion catches the light.
  item: {
    rot: [-0.10, -0.42, 0.20], anchor: [0.66, -0.98], size: 0.36, dist: 0.62,
    grip: [0, 0, 0], flip: false,
  },
};

// Block render shapes that want something other than the cube pose. Keyed on
// the shape rather than the block, so every wood's door gets it for free.
const SHAPE_STYLES = { door: "panel", trapdoor: "panel", ladder: "panel" };

// Which style an item is held in. `mesh` is its entry from the shared item-model
// cache (render/itemmesh.js): `kind` is "shape" for solid blocks and "sprite"
// for everything else, `shape` names a block's render shape.
export function styleFor(item, mesh) {
  if (item && item.hold && STYLES[item.hold]) return STYLES[item.hold];
  if (mesh && mesh.kind === "shape") {
    return STYLES[SHAPE_STYLES[mesh.shape]] || STYLES.block;
  }
  if (item) {
    if (item.toolType === "sword") return STYLES.sword;
    if (item.type === "tool") return STYLES.tool;
    if (item.type === "food") return STYLES.food;
  }
  return STYLES.item;
}

const SWING = 0.26;   // seconds for one swing arc
const EQUIP = 0.18;   // seconds for a newly selected item to rise into frame

export class Viewmodel {
  constructor() {
    this.key = null;      // item key currently in hand
    this.swingT = 1;      // 0..1 through a swing; >= 1 means idle
    this.equipT = 1;      // 0 = fully lowered out of frame, 1 = in place
    this.clock = 0;       // free-running seconds, for the idle drift
    this.bob = null;      // {phase, mag} from the player's walk cycle
    this._queued = false; // a swing asked for while one was already running
  }

  // Called every frame with the selected hotbar key: a change lowers the old
  // item and raises the new one.
  setItem(key) {
    if (key === this.key) return;
    this.key = key;
    this.equipT = 0;
    this.swingT = 1;
    this._queued = false;
  }

  // Start a swing, or queue one so a swing triggered mid-arc still happens.
  // Held mining calls this every frame, which chains into a continuous swing.
  swing() {
    if (this.swingT >= 1) this.swingT = 0;
    else this._queued = true;
  }

  update(dt, bob) {
    this.clock += dt;
    this.bob = bob;
    if (this.equipT < 1) this.equipT = Math.min(1, this.equipT + dt / EQUIP);
    if (this.swingT < 1) {
      this.swingT += dt / SWING;
      if (this.swingT >= 1 && this._queued) { this.swingT = 0; this._queued = false; }
    }
  }

  // Build the view-space model matrix for one item (column-major, for GL).
  // `item` is the registry entry (may be null), `mesh` its item-model entry.
  modelMatrix(out, item, mesh, fovRad, aspect) {
    const st = styleFor(item, mesh);
    let [rx, ry, rz] = st.rot;
    let ax = st.anchor[0], ay = st.anchor[1];

    // --- swing: a chop that drops toward the crosshair, rolls the head over and
    // snaps back. `g` is the broad arc (fast out, slow back), `f` peaks later
    // and harder — the moment of contact. Both are 0 at the ends, so the pose is
    // untouched at rest no matter how the style is tuned.
    if (this.swingT < 1) {
      const t = this.swingT;
      const g = Math.sin(Math.sqrt(t) * Math.PI);
      const f = Math.sin(t * t * Math.PI);
      ax -= 0.13 * g;
      ay -= 0.20 * g;
      rx -= 0.78 * g;
      rz -= 0.26 * g;
      ry += 0.30 * f;
    }

    // --- equip: the item rises from below the frame when you switch slots
    if (this.equipT < 1) {
      const e = 1 - this.equipT;
      ay -= 0.95 * e;
      rx -= 0.35 * e;
      rz += 0.20 * e;
    }

    // --- walk sway: the same bob phase that moves the camera, so the hand and
    // the head agree about which foot is down
    const bob = this.bob;
    if (bob && bob.mag > 0.001) {
      const m = bob.mag;
      ax += Math.cos(bob.phase) * 0.10 * m;
      ay -= Math.abs(Math.sin(bob.phase)) * 0.10 * m;
      rz += Math.sin(bob.phase) * 0.05 * m;
    }

    // --- idle drift: a slow breath so a still hand is never perfectly frozen
    const c = this.clock;
    ax += Math.sin(c * 0.63) * 0.010;
    ay += Math.sin(c * 0.91) * 0.013;
    rz += Math.sin(c * 0.77) * 0.014;

    // --- framing: anchor and size are fractions of the frame at `dist`, so the
    // item keeps its place in the picture at any aspect ratio
    const halfH = st.dist * Math.tan(fovRad / 2);
    const px = ax * halfH * aspect, py = ay * halfH, pz = -st.dist;
    const s = st.size * 2 * halfH;

    // --- compose T(p) . Ry . Rx . Rz . S . T(-grip) . flip.
    // Written out rather than built from four matrix products: this runs every
    // frame for one object, and the expanded form is the whole transform in one
    // place. R = Ry*Rx*Rz, rows below, stored column-major for GL.
    const cx = Math.cos(rx), sx = Math.sin(rx);
    const cy = Math.cos(ry), sy = Math.sin(ry);
    const cz = Math.cos(rz), sz = Math.sin(rz);
    const r00 = cy * cz + sy * sx * sz, r01 = -cy * sz + sy * sx * cz, r02 = sy * cx;
    const r10 = cx * sz,                r11 = cx * cz,                 r12 = -sx;
    const r20 = -sy * cz + cy * sx * sz, r21 = sy * sz + cy * sx * cz, r22 = cy * cx;

    const fx = st.flip ? -s : s;          // column 0 carries the mirror
    out[0] = r00 * fx; out[1] = r10 * fx; out[2] = r20 * fx; out[3] = 0;
    out[4] = r01 * s;  out[5] = r11 * s;  out[6] = r21 * s;  out[7] = 0;
    out[8] = r02 * s;  out[9] = r12 * s;  out[10] = r22 * s; out[11] = 0;
    // The grip is given in MIRRORED model space (what you see), so it moves
    // along the un-mirrored rotated axes — hence r00*s here, not out[0].
    const [gx, gy, gz] = st.grip;
    out[12] = px - (r00 * gx + r01 * gy + r02 * gz) * s;
    out[13] = py - (r10 * gx + r11 * gy + r12 * gz) * s;
    out[14] = pz - (r20 * gx + r21 * gy + r22 * gz) * s;
    out[15] = 1;
    return out;
  }
}

File: src/render/viewmodel.cpp
#include "render/viewmodel.h"

#include <algorithm>
#include <cmath>

namespace hr::render {
namespace {

constexpr float kSwingSeconds = 0.26f;  // one swing arc
constexpr float kEquipSeconds = 0.18f;  // a newly selected item rising into frame

// Ry * Rx * Rz, row-major, matching the convention HoldStyle::rot documents.
struct Rot3 {
  float m[3][3];
};

Rot3 euler(float rx, float ry, float rz) {
  const float cx = std::cos(rx), sx = std::sin(rx);
  const float cy = std::cos(ry), sy = std::sin(ry);
  const float cz = std::cos(rz), sz = std::sin(rz);
  Rot3 r;
  r.m[0][0] = cy * cz + sy * sx * sz;
  r.m[0][1] = -cy * sz + sy * sx * cz;
  r.m[0][2] = sy * cx;
  r.m[1][0] = cx * sz;
  r.m[1][1] = cx * cz;
  r.m[1][2] = -sx;
  r.m[2][0] = -sy * cz + cy * sx * sz;
  r.m[2][1] = sy * sz + cy * sx * cz;
  r.m[2][2] = cy * cx;
  return r;
}

Rot3 operator*(const Rot3& a, const Rot3& b) {
  Rot3 o;
  for (int i = 0; i < 3; ++i) {
    for (int j = 0; j < 3; ++j) {
      o.m[i][j] = a.m[i][0] * b.m[0][j] + a.m[i][1] * b.m[1][j] + a.m[i][2] * b.m[2][j];
    }
  }
  return o;
}

}  // namespace

void Viewmodel::setItem(const std::string& key) {
  if (hasItem_ && key == key_) return;
  key_ = key;
  hasItem_ = true;
  equipT_ = 0.0f;
  swingT_ = 1.0f;
  queued_ = false;
}

void Viewmodel::swing() {
  if (swingT_ >= 1.0f) {
    swingT_ = 0.0f;
  } else {
    queued_ = true;
  }
}

void Viewmodel::cancelSwing() {
  swingT_ = 1.0f;
  queued_ = false;
}

void Viewmodel::update(float dt, float bobPhase, float bobMagnitude) {
  clock_ += dt;
  bobPhase_ = bobPhase;
  bobMagnitude_ = bobMagnitude;
  if (equipT_ < 1.0f) equipT_ = std::min(1.0f, equipT_ + dt / kEquipSeconds);
  if (swingT_ < 1.0f) {
    swingT_ += dt / kSwingSeconds;
    if (swingT_ >= 1.0f && queued_) {
      swingT_ = 0.0f;
      queued_ = false;
    }
  }
}

Mat4 Viewmodel::modelMatrix(const HoldStyle& st, float fovRadians, float aspect) const {
  // The animation is kept OUT of the pose and applied outside it, about the view's
  // own axes: ax swings the item down the screen, ay turns it across the screen,
  // az rolls it in the picture plane.
  //
  // These used to be added straight into the hold style's Euler angles, which is
  // wrong for a reason that stayed hidden while every style pointed roughly the
  // same way: the axis a style's `rot.x` turns about is itself rotated by that
  // style's `rot.y`. Yaw a tool a quarter turn to put its edge into the block —
  // which is what a tool wants — and the "swing down" term becomes a rotation
  // about the view's Z instead, so the tool spun in the picture plane and dug
  // sideways while looking perfectly correct at rest. Held out here, the arm
  // swings down whatever the hand happens to be holding or how it grips it.
  float sx = 0.0f, sy = 0.0f, sz = 0.0f;
  float ax = st.anchorX, ay = st.anchorY;

  // --- swing: a chop that drops toward the crosshair, rolls the head over and
  // snaps back. `g` is the broad arc (fast out, slow back), `f` peaks later and
  // harder — the moment of contact. Both are 0 at the ends, so the pose is
  // untouched at rest no matter how the style is tuned.
  if (swingT_ < 1.0f) {
    const float t = swingT_;
    const float g = std::sin(std::sqrt(t) * 3.14159265358979323846f);
    const float f = std::sin(t * t * 3.14159265358979323846f);
    ax -= 0.13f * g;
    ay -= 0.20f * g;
    sx -= 0.78f * g;  // the head drops toward the crosshair
    sz -= 0.26f * g;  // and leans over as it goes
    sy += 0.30f * f;  // a turn into the blow, at the moment of contact
  }

  // --- equip: the item rises from below the frame when you switch slots.
  if (equipT_ < 1.0f) {
    const float e = 1.0f - equipT_;
    ay -= 0.95f * e;
    sx -= 0.35f * e;
    sz += 0.20f * e;
  }

  // --- walk sway, on the same bob phase that moves the camera.
  if (bobMagnitude_ > 0.001f) {
    const float m = bobMagnitude_;
    ax += std::cos(bobPhase_) * 0.10f * m;
    ay -= std::fabs(std::sin(bobPhase_)) * 0.10f * m;
    sz += std::sin(bobPhase_) * 0.05f * m;
  }

  // --- idle drift: a slow breath, so a still hand is never perfectly frozen.
  ax += std::sin(clock_ * 0.63f) * 0.010f;
  ay += std::sin(clock_ * 0.91f) * 0.013f;
  sz += std::sin(clock_ * 0.77f) * 0.014f;

  // --- framing: anchor and size are fractions of the frame at `dist`, so the item
  // keeps its place in the picture at any aspect ratio.
  const float halfH = st.dist * std::tan(fovRadians / 2.0f);
  const float px = ax * halfH * aspect, py = ay * halfH, pz = -st.dist;
  const float s = st.size * 2.0f * halfH;

  // --- compose T(p) . Rswing . Rpose . S . T(-grip) . flip.
  const Rot3 rot = euler(sx, sy, sz) * euler(st.rot.x, st.rot.y, st.rot.z);
  const float r00 = rot.m[0][0], r01 = rot.m[0][1], r02 = rot.m[0][2];
  const float r10 = rot.m[1][0], r11 = rot.m[1][1], r12 = rot.m[1][2];
  const float r20 = rot.m[2][0], r21 = rot.m[2][1], r22 = rot.m[2][2];

  Mat4 out;
  const float fx = st.flip ? -s : s;  // column 0 carries the mirror
  out.m[0] = r00 * fx;
  out.m[1] = r10 * fx;
  out.m[2] = r20 * fx;
  out.m[3] = 0;
  out.m[4] = r01 * s;
  out.m[5] = r11 * s;
  out.m[6] = r21 * s;
  out.m[7] = 0;
  out.m[8] = r02 * s;
  out.m[9] = r12 * s;
  out.m[10] = r22 * s;
  out.m[11] = 0;
  // The grip is given in MIRRORED model space (what you see), so it moves along the
  // un-mirrored rotated axes — hence r00 * s here, not out.m[0].
  const float gx = st.grip.x, gy = st.grip.y, gz = st.grip.z;
  out.m[12] = px - (r00 * gx + r01 * gy + r02 * gz) * s;
  out.m[13] = py - (r10 * gx + r11 * gy + r12 * gz) * s;
  out.m[14] = pz - (r20 * gx + r21 * gy + r22 * gz) * s;
  out.m[15] = 1;
  return out;
}

}  // namespace hr::render
